src/newtondynamics_physics/SysNewton.cpp

```diff
@@ -0,0 +1,428 @@
+    {
+        // Create world if not yet initialized
+        worldComp->m_nwtWorld = NewtonCreate();
+        NewtonWorldSetUserData(worldComp->m_nwtWorld, &rScene);
+    }
+
+    NewtonWorld const* nwtWorld = worldComp->m_nwtWorld;
+
+    auto viewBody = rScene.get_registry().view<ACompRigidBody>();
+
+    // Iterate all rigid bodies in scene
+    for (ActiveEnt ent : viewBody)
+    {
+        auto &entBody = viewBody.get<ACompRigidBody>(ent);
+
+        // temporary: delete Newton Body if something is dirty
+        if (entBody.m_colliderDirty)
+        {
+            rReg.remove_if_exists<ACompNwtBody>(ent);
+            entBody.m_colliderDirty = false;
+        }
+
+        // Initialize body if not done so yet;
+        if (! rReg.any_of<ACompNwtBody>(ent))
+        {
+            create_body(rScene, ent, nwtWorld);
+        }
+
+        // Recompute inertia and center of mass if rigidbody has been modified
+        if (entBody.m_inertiaDirty)
+        {
+            compute_rigidbody_inertia(rScene, ent);
+            SPDLOG_LOGGER_TRACE(m_scene.get_application().get_logger(), "Updating RB : new CoM Z = {}",
+                              entBody.m_centerOfMassOffset.z());
+        }
+
+    }
+
+    // Update the world
+    NewtonUpdate(nwtWorld, rScene.get_time_delta_fixed());
+
+    auto viewNwtBody = rScene.get_registry().view<ACompNwtBody>();
+    auto viewTf = rScene.get_registry().view<ACompTransform>();
+
+    // Apply transform changes after the Newton world(s) updates
+    for (ActiveEnt ent : viewBody)
+    {
+        auto &entNwtBody      = viewNwtBody.get<ACompNwtBody>(ent);
+        auto &entTransform    = viewTf.get<ACompTransform>(ent);
+
+        float mass, x, y, z;
+        NewtonBodyGetMass(entNwtBody.body(), &mass, &x, &y, &z);
+
+        // Get new transform matrix from newton
+        NewtonBodyGetMatrix(entNwtBody.body(), entTransform.m_transform.data());
+    }
+}
+
+void SysNewton::find_colliders_recurse(
+        ActiveScene& rScene, ActiveEnt ent, Matrix4 const &transform,
+        NewtonWorld const* nwtWorld, NewtonCollision *rCompound)
+{
+    ActiveReg_t &rReg = rScene.get_registry();
+    ActiveEnt nextChild = ent;
+
+    // iterate through siblings of ent
+    while(nextChild != entt::null)
+    {
+        auto const &childHeir = rScene.reg_get<ACompHierarchy>(nextChild);
+        auto const *childTransform = rScene.get_registry().try_get<ACompTransform>(nextChild);
+
+        if (childTransform != nullptr)
+        {
+            Matrix4 childMatrix = transform * childTransform->m_transform;
+
+            if (rReg.all_of<ACompSolidCollider>(nextChild))
+            {
+                auto* childNwtCollider = rReg.try_get<ACompNwtCollider>(nextChild);
+
+                NewtonCollision const *collision;
+
+                if (nullptr == childNwtCollider)
+                {
+                    // No Newton collider exists yet
+                    collision = NewtonCreateSphere(nwtWorld, 0.5f, 0, NULL);
```

Comment:
I get the feeling that allocating / creating these objects "as-needed" when they need to be used, instead of ensuring they are always created at the same time as the component if they are intended to be used later, is a pessimization.

Reply:
It at least ensures that only SysNewton is the one that calls Newton functions. Creating components are cheap, and when Newton functions are called together, there's a good chance all the memory they'll touch will be loaded into the cache less often.

Reply:
I propose making helper functions that hide the fact that functions from newton dynamics are being called, but otherwise do all of the necessary initialization work.

That lets the system that my comment here is inside of just blow through all it's work without needing to check if initialization has been done properly.

Consider adding a check / assert for debug builds though.